main/src/library/MutHeap.flix

```diff
@@ -0,0 +1,240 @@
+
+
+    ///
+    /// Returns the number of elements in a MutHeap
+    ///
+    pub def size(h: MutHeap[a, r]): Int32 \ r =
+        let MutHeap(_, _, s) = h;
+        deref s
+
+    ///
+    /// Returns whether a MutHeap is empty (has 0 elements)
+    ///
+    pub def isEmpty(h: MutHeap[a, r]): Bool \ r =
+        let MutHeap(_, _, s) = h;
+        deref s == 0
+
+    ///
+    /// Returns the top element of a MutHeap
+    ///
+    pub def peek(h: MutHeap[a, r]): Option[a] \ r =
+        let MutHeap(_, arr, s) = h;
+        if (deref s == 0) None
+        else Some(Array.get(0, deref arr))
+
+    ///
+    /// Removes and returns the top element of a MutHeap
+    ///
+    pub def pop(h: MutHeap[a, r]): Option[a] \ r with Order[a] =
+        let MutHeap(_, _, s) = h;
+        if (deref s > 0) {
+            match peek(h) {
+                case Some(elem) => {
+                    delete(h, elem);
+                    Some(elem)
+                }
+                case _          => None
+            }
+        }
+        else None
+
+    ///
+    /// Inserts an element into a MutHeap
+    ///
+    pub def insert(h: MutHeap[a, r], elem: a): Unit \ r with Order[a] = {
+        let MutHeap(_, arr, s) = h;
+        if (deref s == 0) {
+            Array.put(elem, 0, deref arr);
+            s := 1
+        }
+        else {
+            expand!(h);
+            Array.put(elem, deref s, deref arr);
+            heapifyUp(deref s, h);
+            s := deref s + 1
+        }
+    }
+
+    ///
+    /// Deletes an element from a MutHeap
+    ///
+    pub def delete(h: MutHeap[a, r], elem: a): Unit \ r with Order[a] = {
+        let MutHeap(_, arr, s) = h;
+        match Array.findIndexOf(x -> x == elem, deref arr) {
+            case Some(idx) =>
+                let lastElem = Array.get(deref s - 1, deref arr);
+                Array.put(lastElem, idx, deref arr);
+                heapifyDown(idx, h);
+                s := deref s - 1
+            case None => ()
+        }
+    }
+
+    ///
+    /// Returns a string representation of the MutHeap
+    ///
+    pub def toString(h: MutHeap[a, r]): String \ r with ToString[a] = region reg {
+        let MutHeap(_, arr, s) = h;
+        let sb = StringBuilder.new(reg);
+        StringBuilder.appendString!("MutHeap {", sb);
+        Array.forEachWithIndex((i, x) -> {
+            if (i < deref s) {
+                if (i == 0) StringBuilder.appendString!("${x}", sb)
+                else        StringBuilder.appendString!(", ${x}", sb)
+            }
+            else ()
+        }, deref arr);
+        StringBuilder.appendString!("}", sb);
+        StringBuilder.toString(sb)
+    }
+
+    ///
+    /// Returns an iterator over MutHeap `h`
+    ///
+    /// Modifying `h` during iteration is undefined
+    ///
+    pub def iterator(rc: Region[r1], h: MutHeap[a, r2]): Iterator[a, r1 + r2, r1] \ {r1, r2} =
+        let MutHeap(_, arr, s) = h;
+        let numElems = deref s;
+        let idx = ref 0 @ rc;
```

Comment:
I tihnk this could be made simpler by just doing `Iterator.range` and then `Iterator.map`